### PR TITLE
Parser refactor

### PR DESCRIPTION
**Pull request by gdolle**
Major refactoring for the parser. Breaking changes (3.1.0 ?)

Changelog

- Fix options bugs (random behaviour between book.json and local command, ace plugin does not support quotes in command options)
- Add Url support #42 (sync http request using [sync-request](https://www.npmjs.com/package/sync-request) modules + check url [valid-url](https://www.npmjs.com/package/valid-url))
- Refactor options using immutable maps (more robust). It makes easier future new options, see `src/options.js`
- Support real types (not retrocompatible for book options). For example `edit:true` instead of `edit:"true"`,  to be more consistent (this fix an ace plugin bug).
- Add parser checks for types.
- Add different warn level.
- `template`, `fixlang`, `lang`, `theme` are now also local/global options. We can set a template per code, a theme per code, etc.. or globally.
- Refactor test for these last changes.
- Add new tests for URL
- End support of `lang-typescript` to use uniform options `lang:"typescript"` to be consistent with other options and to simplify functions.
- Support multiple snippet codes. Now we can do `[import:"tag0,tag1,..."](fixtures/test.js)` to import several snippets into one codeblock , for example
```cpp
//! [tag0]
int a;
//! [tag0]
// A comment or a code
//! [tag1]
int b;
//! [tag1]
```
will produce
```cpp
int a;
int b;
```

It should work far better with this PR :)

I've been noticed that ace plugin might crash after a page reload in previous version, but I couldn't reproduce that problem. However, these changes fix some bugs and "might" be fix for that behaviour, we will see.

NB: It might be interesting to have a look to async operation in the future to optimize the code.

@azu waiting for your review, you'll have some work ;)

### CODE REVIEW

**Comment by azu** [comment]
Oh, large diff.. 
I'll look it lator.
Can you separate refactor PR and feature PR if possible.

File: src/parser.js [hunk 1]
```diff
@@ -36,65 +42,127 @@ export function splitLabelToCommands(label = "") {
...
+    return immutable.Map(kv);
+}
+
+/**
+ * return content from file or url.
+ * @param {string} filePath
+ * @param {string} originalPath
+ * @return {string}
+ */
+export function getContent( filePath, originalPath )
```
Comment (gdolle) [inline src/parser.js]: Part of feature #42

File: templates/ace-template.hbs [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {% if file.type=="asciidoc" %}++++{% endif %}
-{%ace edit="{{edit}}", check="{{check}}", theme="{{theme}}", lang="{{lang}}" %}
+{%ace edit={{edit}}, check={{check}}, theme="{{theme}}", lang="{{lang}}" %}
```
Comment (gdolle) [inline templates/ace-template.hbs:2]: Major bug fix here

**Comment by gdolle** [comment]
@azu Actually, features are minors and deeply dependent of the of refactor. It'd be quite difficult to separate now. 

Would it be ok to keep it as one PR ?

The related features changes are
- multiple markers (parseVariableFromLabel => change the regex to have several comma ~1line)
- URL parser ( getContent => load url or file ~5lines) 
+ the 1 tests for each in `test-parser.js`

(I added comments on file changes to help)

**Comment by azu** [comment]
@gdolle fmm, I want to separate big PR to smll PRs by following reasons.

> Add Url support #42 (sync http request using sync-request modules + check url valid-url)

Sync request is simple solution, but it is not correct solution.
Because GitBook support async hook.

- https://toolchain.gitbook.com/plugins/hooks.html

(Not actuall code)
```js
// THIS IS IMAGINE
module.exports = {
    hooks: {
        "page:before": function(page) {
            var options = this.options.pluginsConfig["include-codeblock"];
            var pageDir = path.dirname(page.rawPath);
            var parseResults = parse(page.content, pageDir, options);
            var promises = results.map(result => {
                return requestUrlAndCreateReplaceContent(result).then(newContent => {
                     replaceContent(page, newContent);
                });
            });
            return Promise.all(promises); // GitBook support async by returing promise
        }
    }
};
```

and

[Immutable.js](https://facebook.github.io/immutable-js/ "Immutable") is overlarge for this project.
I recommened that use ES2015 [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map "Map") + Poilyfill insteadof immutable.js.

-----

Of course, Welcome to [bugfix](https://github.com/azu/gitbook-plugin-include-codeblock/pull/43#discussion_r97314114), but the bugfix should not contains other feature.
I hesitate to merge the bugfix PR.

Sorry, slow response.

**Comment by gdolle** [comment]
Does ES2015 Map guaranty const Map values ?

Actually refactor parser:
1. immutable Map for read only (more robust to my opinion) 
2. handle key-val types (boolean, string, number). The "bug fix" is a consequence of the parser refactor (maybe my single comment is error prone) it's not really a feature.

I agree for the url feature, it is from far not perfect. I'll separate it in another PR.

File: src/parser.js [hunk 3]
```diff
@@ -36,65 +42,127 @@ export function splitLabelToCommands(label = "") {
...
+export function parseVariablesFromLabel(kvMap,label) {
+    const kv = kvMap.toObject();
+    const begin_ex = "\^.*";
+    const end_ex = ".*\$";
+    const sep_ex = ",?";
+    const kvsep_ex = "[:=]";
+    const spaces_ex = "\\s*";
+    const quotes_ex = "[\"']";
+
+    Object.keys(kv).forEach(key => {
```
Comment (azu) [inline src/parser.js]: Can ES2015 Map does same thing?

```js
const resultMap = new Map();
kvMap.forEach(key => { 
   // something
   resultMap.set(key, foo);
}
return resultMap;
```

File: src/parser.js [hunk 4]
```diff
@@ -103,114 +171,69 @@ export function embedCode({lang, filePath, originalPath, label, template, uninde
...
-
+export function generateEmbedCode(
+    kvMap,
+    {fileName, originalPath, content})
+{   
+    const tContent = getTemplateContent(kvMap);
+    const kv = kvMap.toObject();
+    const count = hasTitle(kv) ? codeCounter() : -1;
+    checkMapTypes( kvMap, "generatedEmbedCode" );
+    const contextMap = kvMap.concat( immutable.Map({
```
Comment (azu) [inline src/parser.js]: `var mergedMap = new Map([...kvMap, ...otherMap])`

```js
var a = new Map([["k1", "v1"], ["k2", "v2"]]);
var b = new Map([["k3", "v3"], ["k4", "v4"]]);
var mergedMap = new Map([...a, ...b]);
console.log(mergedMap); // Map { k1: "v1", k2: "v2", k3: "v3", k4: "v4" }
```


```js
// Object to Map
new Map(Object.entries({ k : 1 }))
```

**Review by azu — COMMENTED** [review]
> Does ES2015 Map guaranty const Map values ?

Yes and No.

Immutable.js is great library, but it is not first class in JavaScript.
In other word, When other constributor see the code that includes immutable.js, they have to learn about immutable.js. I think that it contains cost.

I've written some Map examples.

If you actualy want to use immutable.js, I can't prevent it.

> I agree for the url feature, it is from far not perfect. I'll separate it in another PR.

Ok.

**Comment by azu** [comment]
(Currently, this project mixed codeing style, We should apply ESLint  #39)

**Comment by gdolle** [comment]
@azu I updated the code. 
- Remove immutable: I used object instead of Map and froze object that should be immutable. It should be understandable for future contributors.
- Remove URL feature.

I didn't apply yet Eslint, I'll have a look into that.

**Comment by gdolle** [comment]
@azu I added eslint config following #39 and fix current code following the convention.
I also added a "check" script which test eslint on src/ and test/.
This script is executed in travis, therefore we can see if something is not good with respect to the coding style.
Let me know if it's ok for you like that. ;)

File: src/parser.js [hunk 5]
```diff
@@ -36,181 +38,176 @@ export function splitLabelToCommands(label = "") {
...
+/**
+ * Unindent code
+ * @param {string} s
+ * @return {string}
+ */
+export function strip(s) {
+  // inspired from https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/string/strip.rb
+  if((s === undefined) || (s === "")) {
+return s;
+}
```
Comment (azu) [inline src/parser.js]: Can you format indent or Run `$(npm bin)/eslint --fix src/ test/`?

File: src/parser.js [hunk 6]
```diff
@@ -36,181 +38,176 @@ export function splitLabelToCommands(label = "") {
...
+            case "string":
+                valEx = quotesEx + valEx + quotesEx;
+                break;
+            case "boolean":
+                // no quotes
+                valEx = quotesEx + "?(true|false)" + quotesEx + "?";
+                break;
+            default:
+                logger.error("include-codeblock: parseVariablesFromLabel: key type `" +
+                    typeof defaultKeyValueMap[key] + "` unknown (see options.js)");
```
Comment (azu) [inline src/parser.js:103]: add `breaK;` to `default:` block.

File: src/parser.js [hunk 7]
```diff
@@ -36,181 +38,176 @@ export function splitLabelToCommands(label = "") {
...
-    else if (hasMarker(keyValueObject)) {
-        const marker = getMarker(keyValueObject);
+    } else if (hasMarker(kvm)) {
+        // Slice content via markers.
+        const marker = getMarker(kvm);
         content = removeMarkers(markerSliceCode(code, marker));
     }
-    if (unindent || keyValueObject.unindent) {
-      content = strip(content);
+    if (unindent == true) {
```
Comment (azu) [inline src/parser.js]: use `===` insteadof `==`.

OR

```js
const unindent = kvm.unindent !== undefined ? kvm.unindent : false;
...
if (unindent) { ... }
...
```

File: package.json [hunk 8]
```diff
@@ -11,7 +11,7 @@
   "bugs": {
     "url": "https://github.com/azu/gitbook-plugin-include-codeblock/issues"
   },
-  "version": "3.0.1",
+  "version": "3.1.0",
```
Comment (azu) [inline package.json]: PR is not include bump version.

File: package.json [hunk 9]
```diff
@@ -24,6 +24,7 @@
   },
   "scripts": {
     "build": "NODE_ENV=production babel src --out-dir lib --source-maps",
+    "check": "eslint src/*.js test/*.js",
     "watch": "babel src --out-dir lib --watch --source-maps",
```
Comment (azu) [inline package.json]: Maybe, `lint` command is meaningful.

Also, To define `"lint:fix": "eslint --fix src/*.js test/*.js"` is useful.

**Review by azu — CHANGES_REQUESTED** [review]
Thanks.
I've leave some comments

File: src/parser.js (hunk 5 above)
Comment (gdolle) [reply src/parser.js]: weird eslint does not fix this lines..

**Comment by gdolle** [comment]
Sorry for the late reply. I updated the code relatively to your last review

**Review by azu — APPROVED** [review]

**Comment by azu** [comment]
@gdolle Thanks for great work!

**Pull request merged**